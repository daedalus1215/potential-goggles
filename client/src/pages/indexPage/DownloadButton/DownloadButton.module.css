.buttonDownload,
.buttonDownload:hover {
  background-color: #3e80ed;
  color: white;
  width: 45px;
  height: 40px;
  margin: 4px 5px;
  width: 150px;
}

.title {
  margin-left: 1rem;
}

.icon {
  font-size: 1.2rem;
}


.buttonDownload:hover {
  background-color: #184693;
}