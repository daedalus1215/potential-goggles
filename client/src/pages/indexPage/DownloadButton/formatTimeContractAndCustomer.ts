import { Task } from '../../../interfaces';
import { displayMsInFractionalHourFormat } from '../../../utils';

/**
 * Take a task object and an array of projects and reformat the
 * task object something that can be aggregated and written into
 * a json file.
 *
 * @param {Object} task: a Task object
 * @param {Array} projects: array of projects.
 * @returns {Object} a task that has a format for a json file.
 */
export default function formatTimeContractAndCustomer(task: Task, projects: any) {
  const taskWithProject = { ...task };

  taskWithProject.time = displayMsInFractionalHourFormat(task.time) as any;

  projects.forEach((project: any) => {
    if (taskWithProject.projectId === project.key) {
      taskWithProject.projectId = project.contract;
      taskWithProject.projectId = project.customer;
    }
  });

  return taskWithProject;
}
