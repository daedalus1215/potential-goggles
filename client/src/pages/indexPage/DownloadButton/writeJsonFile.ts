import FileSaver from "file-saver";

/**
 * Not exporting default, so we can spyOn
 * @param {Object} taskBundle
 */
export default function writeJsonFile(taskBundle:any) {
  // console.log(taskBundle)
  
  // window.console.log('writeable', taskBundle)
  let json = JSON.stringify(taskBundle);
  let blob = new Blob([json], { type: 'application/json' });
  let fileName = 'time-logs_' + new Date() + '.json';
  FileSaver.saveAs(blob, fileName);
}
