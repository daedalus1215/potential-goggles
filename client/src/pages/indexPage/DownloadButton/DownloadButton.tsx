import React from 'react';
import IconButton from '@/components/iconButton/IconButton';
import writeJsonFile from './writeJsonFile';
import { fetchTasks } from '@/actionsAndLoaders/loaders';

const DownloadButton: React.FC = () => {
  const handleDownload = async () => {
    const tasks = await fetchTasks();
    return writeJsonFile(tasks)
  }

  return <IconButton
    category='primary2nd'
    onClick={handleDownload}
    icon="bi bi-cloud-download"
    title="Download"
  />
};

export default DownloadButton;
