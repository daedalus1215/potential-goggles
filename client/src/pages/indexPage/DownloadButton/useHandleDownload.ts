import writeJsonFile from './writeJsonFile';
import { Task } from '../../../interfaces';
import { displayMsInFractionalHourFormat } from '../../../utils';
import { fetchTasks } from '../../../actionsAndLoaders/actions';

const useHandleDownload = async () => {
    const tasks = await fetchTasks();

    const assembleTask = tasks?.map(task =>
        ({ ...task, time: displayMsInFractionalHourFormat(task.time) })
    );
    return (e?:any) => {
        writeJsonFile(assembleTask);
    };
};

export default useHandleDownload;