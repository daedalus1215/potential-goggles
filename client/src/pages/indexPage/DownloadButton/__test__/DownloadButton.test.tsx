import { render } from '@testing-library/react';
import '@testing-library/jest-dom'

import DownloadButton from '../DownloadButton';

describe('src/pages/indexPage/DownloadButton/__test__/DownloadButton.test.tsx', () => {
    describe('DownloadButton', () => {
        it('should render', () => {
            // Arrange & Act
            const target = render(<DownloadButton />);

            // Assert
            expect(target.queryByTestId('icon-button')).toBeInTheDocument();
            expect(target.queryByTestId('icon-button')).toMatchSnapshot();
        });
    });
});