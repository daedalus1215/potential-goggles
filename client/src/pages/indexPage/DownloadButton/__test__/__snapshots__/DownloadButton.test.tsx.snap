// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/indexPage/DownloadButton/__test__/DownloadButton.test.tsx DownloadButton should render 1`] = `
<button
  class="baseBtn IconButton hasTitle primary2nd"
  data-testid="icon-button"
>
  <i
    class="bi bi-cloud-download icon"
  />
  <span
    class="title"
  >
    Download
  </span>
</button>
`;
